### DISCUSSION TITLE
CPU Consumption High with Text Overlay

### DISCUSSION THREAD

**Discussion by BusterNeece** [savonet/liquidsoap#2324] [User Support]
Hello! AzuraCast team here!

Many of our users like to run lightweight video streams that are just their radio streams with a simple looping background video and the Now Playing data overlaid on top of it. Previously, this required a separate ffmpeg container running its own script, which was complicated and quite difficult to keep up and running smoothly. With the release of Liquidsoap 2.0.0, we've been looking to move this entirely into AzuraCast itself, as a custom addition to our pre-generated Liquidsoap configuration.

Here's an example script we're adding onto the end of our radio broadcast (which only ingests a single variable named `radio` that is the composited radio signal):

```
# Station Base Directory
station_base_dir = "/var/azuracast/stations/funkywayfm"

# YouTube Stream URL
url = ""

# Fetch and encode video
video_file = station_base_dir ^ "/media/videostream/video.mp4"

# Write now playing to the video file
font_file = station_base_dir ^ "/media/videostream/font.ttf"

# Text specifications
font_size = 40
font_x = 225
font_y = 650

# Add Now Playing metadata
nowplaying = ref("")
def update_nowplaying(m)
    nowplaying := m["artist"] ^ " - " ^ m["title"]
end
radio.on_metadata(update_nowplaying)

# Method to overlay now playing text
def add_nowplaying_text(d, s) =
  v = ffmpeg.raw.encode.video(%ffmpeg(%video.raw), s)
  def mkfilter(graph)
    v = video.add_text.ffmpeg.raw(font=font_file,size=font_size,speed=0,x=font_x,y=font_y,graph=graph,d,v)
    ffmpeg.filter.video.output(graph, v)
  end
  v = ffmpeg.filter.create(mkfilter)
  s = ffmpeg.raw.decode.video(v)
  s
end

videostream = single(video_file)

videostream = add_nowplaying_text({!nowplaying}, videostream)

# Re-encode the video and audio streams
videostream = ffmpeg.encode.video(
    %ffmpeg(%video(codec="libx264", pixel_format="yuv420p", b="300k", preset="superfast", r=25, g=50)), 
    videostream
)

radio = ffmpeg.encode.audio(
    %ffmpeg(%audio(codec="libmp3lame", samplerate=44100, b="160k")),
    radio
)

# Add the audio stream
videostream = mux_video(video=videostream, radio)

# Output to YouTube
enc = %ffmpeg(
    format="flv", 
    %video.copy,
    %audio.copy
)

output.url(fallible=true, url=url, enc, videostream)
```

This seems to work fine, except there's one problem: this script raw-encodes the video, applies the text filter, then re-decodes it, only to then re-encode it for the output video stream.

This is basically causing twice the CPU load for the same functionality, which pushes what our installation is capable of in this case.

Is there a way to rewrite this script so that it does the raw encode, applies the filter, and then encodes the final broadcast signal from that without any unnecessary transcoding happening?

Thank you!

**Comment by toots** [comment]
Hi,

You should be able to do this whole thing staying on the ffmpeg side of things without going into liquidsoap's internal data. 

Liquidsoap should be able to decode the video single straight into the raw ffmpeg format, pass it to the filter and then to the encoder. You should also be able to mux it with internal format and encode one side as `%audio` and the other as `%video.raw`:

```ruby
# Station Base Directory
station_base_dir = "/var/azuracast/stations/funkywayfm"

# YouTube Stream URL
url = ""

# Fetch and encode video
video_file = station_base_dir ^ "/media/videostream/video.mp4"

# Write now playing to the video file
font_file = station_base_dir ^ "/media/videostream/font.ttf"

# Text specifications
font_size = 40
font_x = 225
font_y = 650

# Add Now Playing metadata
nowplaying = ref("")
def update_nowplaying(m)
    nowplaying := m["artist"] ^ " - " ^ m["title"]
end
radio.on_metadata(update_nowplaying)

# Method to overlay now playing text
def add_nowplaying_text(d, s) =
  def mkfilter(graph) =
    s = video.add_text.ffmpeg.raw(font=font_file,size=font_size,speed=0,x=font_x,y=font_y,graph=graph,d,s)
    ffmpeg.filter.video.output(graph, s)
  end
  ffmpeg.filter.create(mkfilter)
end

videostream = single(video_file)

videostream = add_nowplaying_text({!nowplaying}, videostream)

# Add the audio stream
videostream = mux_video(video=videostream, radio)

# Output to YouTube
enc = %ffmpeg(
    format="flv", 
    %video.raw(codec="libx264", pixel_format="yuv420p", b="300k", preset="superfast", r=25, g=50),
    %audio(codec="libmp3lame", samplerate=44100, b="160k")
)

output.url(fallible=true, url=url, enc, videostream)
```

**Reply by simpsss** [reply]
Hi toots,

My server RAM increased from 2GB to 16GB after this setting, and liquidsoap restarted.

https://i.imgur.com/fZncTp9.png

kern  :err   : [  +0.000143] Out of memory: Killed process 165857 (liquidsoap) total-vm:16368500kB, anon-rss:14638268kB, file-rss:0kB, shmem-rss:0kB, UID:1000 pgtables:30364kB oom_score_adj:0
kern  :info  : [  +0.706445] oom_reaper: reaped process 165857 (liquidsoap), now anon-rss:16kB, file-rss:0kB, shmem-rss:0kB

It works normally in SlvrEagle23's settings.

**Reply by toots** [reply]
Woo thanks for pointing that out. I was able to trace to another case of the OCaml GC not knowing that OCaml values pointing to ffmpeg frames were holding a lot of data and being lazy when cleaning them up. I have published `ocaml-ffmpeg` version `1.1.3` that fixes that.

**Reply by BusterNeece** [reply]
@toots Thanks for the quick turnaround on the fix there. I'm occasionally seeing the whole thing crash, and it seems like it's trying to parse the Now Playing metadata incorrectly and that's triggering it?

```
[drawtext_0 @ 0x55755a18ba80] Missing key or no key/value separator found after key 'Where Cuddles Are Free'
[drawtext_0 @ 0x55755a18ba80] Failed to process command. Continuing with existing parameters.
2022/04/07 09:41:48 [clock.ffmpeg.filter:2] Source buffer_0 failed while streaming: Lang.Runtime_error { kind: "ffmpeg.filter", msg: Some "Avutil.Error(Invalid argument)\nBacktrace:\nRaised by primitive operation at Avfilter.process_command in file \"avfilter/avfilter.ml\", line 287, characters 2-55\nCalled from Builtins_ffmpeg_filters.apply_filter.(fun) in file \"lang/builtins_ffmpeg_filters.ml\", line 313, characters 30-80\n", pos: [at /var/azuracast/.opam/4.13.1/lib/liquidsoap/share/liquidsoap/2.0.4/libs/ffmpeg.liq, line 108 char 2 - line 116 char 5, at /var/azuracast/.opam/4.13.1/lib/liquidsoap/share/liquidsoap/2.0.4/libs/ffmpeg.liq, line 114, char 83-83, at /var/azuracast/.opam/4.13.1/lib/liquidsoap/share/liquidsoap/2.0.4/libs/ffmpeg.liq, line 112 char 17 - line 114 char 83, at /var/azuracast/.opam/4.13.1/lib/liquidsoap/share/liquidsoap/2.0.4/libs/list.liq, line 77 char 0 - line 105 char 16, at /var/azuracast/.opam/4.13.1/lib/liquidsoap/share/liquidsoap/2.0.4/libs/list.liq, line 80 char 2 - line 105 char 16, at /var/azuracast/.opam/4.13.1/lib/liquidsoap/share/liquidsoap/2.0.4/libs/list.liq, line 81, char 21-64, at /var/azuracast/.opam/4.13.1/lib/liquidsoap/share/liquidsoap/2.0.4/libs/list.liq, line 81, char 21-64, at /var/azuracast/.opam/4.13.1/lib/liquidsoap/share/liquidsoap/2.0.4/libs/ffmpeg.liq, line 113 char 17 - line 95 char 3, at /var/azuracast/.opam/4.13.1/lib/liquidsoap/share/liquidsoap/2.0.4/libs/ffmpeg.liq, line 93, char 18-81] }!
2022/04/07 09:41:48 [main:3] Shutdown started!
```

I looked into it and the track title there is "Zodiac: Where Cuddles are Free"...so it looks like it's parsing the colon as part of the command to draw text, even though it's all part of a single string. Any way we can get that looked into?

**Reply by toots** [reply]
Woof, that's another good one, thanks @SlvrEagle23 ! Should be fixed in both `main` and `v2.0.4-preview` now..

**Reply by BusterNeece** [reply]
@toots Running stably now for several hours at a pretty nominal CPU consumption. Will continue to observe and report back if anything changes. Thank you so much for your help and quick response!

**Reply by toots** [reply]
Sweet. Thanks for the feedback y'all this is exactly the kind of stuff we need to keep moving! Planning on releasing `2.0.4` next week. Got only a couple of issues left to look at.
